superbench/benchmarks/micro_benchmarks/ib_validation_performance/ib_validation_performance.cc

```diff
@@ -0,0 +1,450 @@
+    } catch (const std::exception &e) {
+        std::cout << "Error: failed to parse the config, message: " << e.what() << std::endl;
+        MPI_Finalize();
+    }
+
+    return run_in_total;
+}
+
+// Execute shell cmd in termial and get output
+std::string exec(const char *cmd) {
+    char buffer[128];
+    std::string result = "";
+    // use pipe to execute command
+    FILE *pipe = popen(cmd, "r");
+    if (!pipe)
+        throw std::runtime_error("popen() failed!");
+    try {
+        // use buffer to get output
+        while (fgets(buffer, sizeof buffer, pipe) != NULL) {
+            result += buffer;
+        }
+    } catch (const std::exception &e) {
+        pclose(pipe);
+        throw e;
+    }
+    pclose(pipe);
+    return result;
+}
+
+// Get private ip of current machine
+std::string get_my_ip() {
+    std::string ip;
+    try {
+        string command = "hostname -I | awk '{print $1}'";
+        ip = exec(command.c_str());
+        int r = ip.find('\n');
+        ip = ip.substr(0, r);
+    } catch (const std::exception &e) {
+        std::cout << "Error: failed to get ip of " << processor_name << std::endl;
+        MPI_Finalize();
+    }
+    return ip;
+}
+
+// Use socket to get a free port
+int get_available_port() {
+    struct sockaddr_in server_addr;
+    int sockfd;
+    // open a socket
+    if ((sockfd = socket(AF_INET, SOCK_STREAM, 0)) < 0) {
+        std::cout << "Error: failed to open socket.";
+        MPI_Finalize();
+    }
+    server_addr.sin_family = AF_INET;
+    server_addr.sin_addr.s_addr = INADDR_ANY;
+    server_addr.sin_port = 0;
+    inet_aton("127.0.0.1", &server_addr.sin_addr);
+    bzero((char *)&server_addr, sizeof(server_addr));
+    // assign a port number
+    if (bind(sockfd, (struct sockaddr *)&server_addr, sizeof(server_addr)) < 0) {
+        std::cout << "Error: failed to bind socker.";
+        MPI_Finalize();
+    }
+    socklen_t len_inet = sizeof(server_addr);
+    // get socket name
+    if (getsockname(sockfd, (struct sockaddr *)&server_addr, &len_inet) < 0) {
+        std::cout << "Error: failed to get socketname." << endl;
+        MPI_Finalize();
+    }
+    // get the port number
+    int local_port = ntohs(server_addr.sin_port);
+    // make the port can be reused
+    int enable = 1;
+    if (setsockopt(sockfd, SOL_SOCKET, SO_REUSEADDR, &enable, sizeof(int)) < 0) {
+        std::cout << "Error: failed to set socket." << endl;
+        MPI_Finalize();
+    }
+    return local_port;
+}
+
+// Get and broadcast ports for each run
+vector<int> prepare_ports(const vector<std::pair<int, int>> &run_pairs_in_parallel) {
+    int pair_count = run_pairs_in_parallel.size();
+    vector<int> ports(pair_count);
+
+    for (int index = 0; index < run_pairs_in_parallel.size(); index++) {
+        int server_index = run_pairs_in_parallel[index].first;
+        // server get a free port and send to rank 0
+        if (server_index == world_rank) {
+            int port = get_available_port();
```

Comment:
Do we really need this? We can always use a pre-defined port (specific for each device).

And, if we want to do this dynamic, we can exchange the base port as part of initial MPI_AllGather.